drivers/can/stm32_can.c

```diff
@@ -838,51 +838,66 @@ static const struct can_driver_api can_api_funcs = {
 	.detach = can_stm32_detach
 };
 
-#ifdef CONFIG_CAN_1
-
-static void config_can_1_irq(CAN_TypeDef *can);
-
-static const struct can_stm32_config can_stm32_cfg_1 = {
-	.can = (CAN_TypeDef *)CONFIG_CAN_1_BASE_ADDRESS,
-	.bus_speed = CONFIG_CAN_1_BUS_SPEED,
-	.swj = CONFIG_CAN_1_SJW,
-	.prop_bs1 = CONFIG_CAN_1_PROP_SEG_PHASE_SEG1,
-	.bs2 = CONFIG_CAN_1_PHASE_SEG2,
-	.pclken = {
-		.enr = CONFIG_CAN_1_CLOCK_BITS,
-		.bus = CONFIG_CAN_1_CLOCK_BUS,
-	},
-	.config_irq = config_can_1_irq
-};
-
-static struct can_stm32_data can_stm32_dev_data_1;
-
-DEVICE_AND_API_INIT(can_stm32_1, CONFIG_CAN_1_NAME, &can_stm32_init,
-		    &can_stm32_dev_data_1, &can_stm32_cfg_1,
-		    POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
-		    &can_api_funcs);
-
-static void config_can_1_irq(CAN_TypeDef *can)
-{
-	SYS_LOG_DBG("Enable CAN1 IRQ");
-#ifdef CONFIG_SOC_SERIES_STM32F0X
-	IRQ_CONNECT(CONFIG_CAN_1_IRQ, CONFIG_CAN_1_IRQ_PRIORITY, can_stm32_isr,
-		    DEVICE_GET(can_stm32_1), 0);
-	irq_enable(CONFIG_CAN_1_IRQ);
-#else
-	IRQ_CONNECT(CONFIG_CAN_1_IRQ_RX0, CONFIG_CAN_1_IRQ_PRIORITY,
-		    can_stm32_rx_isr, DEVICE_GET(can_stm32_1), 0);
-	irq_enable(CONFIG_CAN_1_IRQ_RX0);
-
-	IRQ_CONNECT(CONFIG_CAN_1_IRQ_TX, CONFIG_CAN_1_IRQ_PRIORITY,
-		    can_stm32_tx_isr, DEVICE_GET(can_stm32_1), 0);
-	irq_enable(CONFIG_CAN_1_IRQ_TX);
-
-	IRQ_CONNECT(CONFIG_CAN_1_IRQ_SCE, CONFIG_CAN_1_IRQ_PRIORITY,
-		    can_stm32_tx_isr, DEVICE_GET(can_stm32_1), 0);
-	irq_enable(CONFIG_CAN_1_IRQ_SCE);
-#endif
-	can->IER |= CAN_IT_TME | CAN_IT_ERR | CAN_IT_FMP0 | CAN_IT_FMP1;
-}
-
-#endif /*CONFIG_CAN_1*/
+/**
+ * @code{.codegen}
+ * codegen.import_module('devicedeclare')
+ *
+ * device_configs = ['CONFIG_CAN_1', ]
+ * driver_names = ['CAN_1', ]
+ * device_inits = 'can_stm32_init'
+ * device_levels = 'POST_KERNEL'
+ * device_prios = 'CONFIG_KERNEL_INIT_PRIORITY_DEVICE'
+ * device_api = 'can_api_funcs'
+ * device_info = \
+ * """
+ * DEVICE_DECLARE(${device-name});
+ * static void ${device-config-irq}(CAN_TypeDef *can)
+ * {
+ *         SYS_LOG_DBG("Enable ${driver-name} IRQ");
+ * #ifdef CONFIG_SOC_SERIES_STM32F0X
+ *         IRQ_CONNECT(${interrupts/0/irq}, ${interrupts/0/priority},\\
+ *                     can_stm32_isr,
+ *                     DEVICE_GET(${device-name}), 0);
+ *         irq_enable(${interrupts/0/irq});
+ * #else
```

Comment:
@erwango 

This is the raw version of device declare as I explained in several places. Your sophisticated version of device_declare can go without #ifdefs. The raw version shall just show how easily the non codegen code can be transformed to use codegen, and how readability and maintainability is given.

I transfered your device_declare API to the devicedeclare codegen module. Please see how this uses the raw API to provide the sopisticated API.

Reply:
> I think #ifdef should not be available in driver instantiation template (and this is why I'm not in favor of full 'raw' templates).

In your sopisticated device_declare API you are using #ifdef behind the scenes for the interrupt flag. The raw version sould support to be used to implement your more sophisticated API. Also the raw version is only for cases where the sophisticated device declare API does not provide enough control on the driver info template.  Adding any limitations to the usage of C code in the template would render the extended template control worthless.

> So, either generation will break because properties are not available.

Generation will not break - but compilation will break if DTS properties are not available because the placeholder will not be replaced in this case.

> And it also means you're not able to generate a warning/error when a missing property occurs in the F0 case (which is, I think, required).

Compiler error should be enough to provide guidance for the driver writer/ tester. There is no undetected inavlid code.